### ISSUE TITLE
Ability to load dependency model from non-app assembly

### ISSUE THREAD

**Issue by piotrpMSFT** [dotnet/core-setup#625]
Moved from https://github.com/dotnet/cli/issues/3253 on behalf of @pakrym.

----

To allow custom compilation assembly load locations.

@pranavkm @davidfowl

**Comment by danroth27** [comment]
Renaming to reflect what we really want here.

**Comment by muratg** [comment]
Assigning to @gkhanna79 and @eerhardt based on our discussion last week.

@gkhanna79 This is not a scenario we care about from MVC point of view in 2.0.0. Please feel free to move it to a future milestone.

**Comment by eerhardt** [comment]
@danroth27 - I think your rename has drastically changed the intention of this issue.

The original issue was asking to "allow for custom compilation assembly load locations", meaning to be able to add more https://github.com/dotnet/core-setup/search?utf8=%E2%9C%93&q=ICompilationAssemblyResolver instances.

I'm not sure what the new name is asking for.  Do you want to retrieve the DependencyModel for a non-app assembly (i.e. `LoadDependencyModel("SomeLibrary.dll")`)?  Or do you want to be able to retrieve the DependencyModel of the current application from code within some non-app assembly/library?

Either way, I think all 3 of these things are distinctly different.

**Comment by danroth27** [comment]
@davidfowl 

I want to be able to get the DependencyModel for a different assembly than the current app assembly. This came up when I was working on ASP.NET Core modules, where I wanted to be able to add a package to my app that brings in MVC controllers and views. The discovery of controllers for that package would need to be scoped to the dependency model for just that package.

**Comment by muratg** [comment]
cc @Eilon

**Comment by davidfowl** [comment]
AFAIK the issue is with the fact that we only search the application base for dependencies.

/cc @pakrym

**Comment by pakrym** [comment]
Yes, we can get dependency model for random assembly but we are not using its location during compilation assembly resolution.

**Comment by eerhardt** [comment]
I think I'm getting more confused as the discussion goes on.

Can we start over and list the exact scenario that doesn't work?  Maybe some repro steps and expected/actual results would help explain.

**Comment by pakrym** [comment]
Scenario in which people hit this issue looks the following way.
There is an application that uses MVC, and a plugin application that user want's to load dynamically from random directory on disk that brings own dependencies, MVC views, ref assemblies.

When doing `DependencyContext.Load` for plugin.dll you compilation library resolution would no work because it donesn't know to search for compilation assemblies relative to plugin.dll and always does it relative to entry assembly.

**Comment by eerhardt** [comment]
@pakrym - let me know if you think https://github.com/eerhardt/core-setup/commit/71bde9e9c04cb5b2355bec6ec843959d54d402e1 will address this issue.  If it will, I'll add tests and send it out for a PR.

**Comment by pakrym** [comment]
@eerhardt It's a start, I'm sure some other issues will rise when people start using DependencyModel this way.

**Comment by eerhardt** [comment]
Well, let me know if this is something you want to pursue. If it is, I can make the change. If it isn't, then we can move this bug out of 2.0.

**Comment by eerhardt** [comment]
It's been a week and no reply.  Moving out of 2.0.

**Comment by danroth27** [comment]
I'd still like to see this go into 2.0. I hit this issue as part of the ASP.NET Core Modules prototyping work and I still think we want this. Right now things are pretty crazy getting basic stuff working for BUILD, but can we revisit this in May?

**Comment by eerhardt** [comment]
I've submitted a PR for the initial ask of "Add ability to customize resolver settings in DependencyModel".

https://github.com/dotnet/core-setup/pull/2154

Since the scenario outlined in [@pakrym's comment](https://github.com/dotnet/core-setup/issues/625#issuecomment-288145796) isn't a .NET Core 2.0 runtime scenario, I can't guarantee we will fix any other bugs that come up in this scenario - especially considering that ZBB is only a few weeks away. But since this small ask seemed reasonable, I went ahead and implemented it.
